include/boost/math/special_functions/jacobi_theta.hpp

```diff
@@ -0,0 +1,616 @@
+    RealType result = RealType(0);
+
+    // n>=0
+    result += _jacobi_theta_sum(tau, z + constants::half_pi<RealType>(), constants::pi<RealType>(), eps);
+    // n<0
+    result += _jacobi_theta_sum(tau, z - constants::half_pi<RealType>(), -constants::pi<RealType>(), eps);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta3tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType result = 0;
+
+    // n=0
+    result += exp(-z*z*tau/constants::pi<RealType>());
+    // n>0
+    result += _jacobi_theta_sum(tau, z + constants::pi<RealType>(), constants::pi<RealType>(), eps);
+    // n<0
+    result += _jacobi_theta_sum(tau, z - constants::pi<RealType>(), -constants::pi<RealType>(), eps);
+
+    return result * sqrt(tau);
+}
+
+template <class RealType, class Policy>
+inline RealType
+_IMAGINARY_jacobi_theta4tau(RealType z, RealType tau, const Policy& pol) {
+    BOOST_MATH_STD_USING
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType result = 0;
+
+    // n = 0
+    result += exp(-z*z*tau/constants::pi<RealType>());
+
+    // n > 0 odd
+    result -= _jacobi_theta_sum(tau, z + constants::pi<RealType>(), constants::two_pi<RealType>(), eps);
+    // n < 0 odd
+    result -= _jacobi_theta_sum(tau, z - constants::pi<RealType>(), -constants::two_pi<RealType>(), eps);
+    // n > 0 even
+    result += _jacobi_theta_sum(tau, z + constants::two_pi<RealType>(), constants::two_pi<RealType>(), eps);
+    // n < 0 even
+    result += _jacobi_theta_sum(tau, z - constants::two_pi<RealType>(), -constants::two_pi<RealType>(), eps);
+
+    return result * sqrt(tau);
+}
+
+// Begin public API
+
+// First Jacobi theta function (Parameterized by tau - assumed imaginary)
+// = 2 * Σ (-1)^n * exp(iπτ*(n+1/2)^2) * sin((2n+1)z)
+template <class RealType, class Policy>
+inline RealType
+jacobi_theta1tau(RealType z, RealType tau, const Policy& pol)
+{
+    BOOST_MATH_STD_USING
+    unsigned n = 0;
+    RealType eps = policies::get_epsilon<RealType, Policy>();
+    RealType q_n, delta, result = RealType(0);
+
+    if (tau <= 0.0)
+        return policies::raise_domain_error<RealType>("boost::math::jacobi_theta1tau<%1%>(%1%)",
+                "tau must be greater than 0 but got %1%.", tau, pol);
+
+    if (abs(z) == 0.0)
+        return result;
+
+    if (tau < 1.0) {
+        z = fmod(z, constants::two_pi<RealType>());
+        while (z > constants::pi<RealType>()) {
+            z -= constants::two_pi<RealType>();
+        }
+        while (z < -constants::pi<RealType>()) {
+            z += constants::two_pi<RealType>();
+        }
+
+        return _IMAGINARY_jacobi_theta1tau(z, 1/tau, pol);
+    }
+
+    do {
+        q_n = exp(-tau * constants::pi<RealType>() * RealType(n + 0.5)*RealType(n + 0.5) );
+        delta = q_n * sin(RealType(2*n+1)*z);
```

Comment:
Calling `sin` on each iteration seems expensive; can't it be done via a Clenshaw recurrence?

Reply:
The iteration count is typically between 2 and 5 so I'm not sure a recurrence formula will buy much (also because there's already an `exp` in each iteration).

Reply:
Ok, so most Clenshaw recurrence discussions descend rather than ascend, so we instead can use

sin((2(n+1)+1)x) = sin((2n+1)x)*cos(2x) + cos((2n+1)x)*sin(2x)

Reply:
Combine this with the recurrence for q^n^2 and you can do the whole thing with a single call to std::cos and std::sin and then only muls and adds after that.